src/rust/engine/src/nodes.rs

```diff
@@ -423,34 +437,80 @@ impl ExecuteProcess {
       timeout: Duration::from_millis((timeout_in_seconds * 1000.0) as u64),
       description: description,
       jdk_home: jdk_home,
-    }))
+    })
+  }
+
+  fn lift(value: &Value) -> Result<MultiPlatformExecuteProcess, String> {
+    let mut request_by_constraint: BTreeMap<(Platform, Platform), ExecuteProcessRequest> =
+      BTreeMap::new();
+    let constraint_parts = externs::project_multi_strs(&value, "platform_constraints");
+    let mut requests: Vec<ExecuteProcessRequest> = Vec::new();
+    for execute_process in externs::project_multi(&value, "execute_process_requests").iter() {
+      let underlying_req = MultiPlatformExecuteProcess::lift_execute_process(execute_process)?;
+      requests.push(underlying_req);
+    }
+    if constraint_parts.len() % 2 != 0 {
+      return Err("Error parsing platform_constraints: odd number of parts".to_owned());
+    }
+    if constraint_parts.len() / 2 != requests.len() {
+      return Err(format!(
+        "Size of constraint keys and requests does not match: {} vs. {}",
+        constraint_parts.len() / 2,
+        requests.len()
+      ));
+    }
+    for (constraint_key_pair, request) in constraint_parts.chunks_exact(2).zip(requests.iter()) {
+      let mut platform_constaint: Vec<Platform> = constraint_key_pair
+        .iter()
+        .map(|c| MultiPlatformExecuteProcess::get_platform_variant(c.to_string()))
+        .collect();
+      request_by_constraint.insert(
+        (platform_constaint.remove(0), platform_constaint.remove(0)),
+        request.clone(),
+      );
+    }
+    Ok(MultiPlatformExecuteProcess(
+      MultiPlatformExecuteProcessRequest(request_by_constraint),
+    ))
   }
 }
 
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct ProcessResult(process_execution::FallibleExecuteProcessResult);
+impl From<MultiPlatformExecuteProcess> for NodeKey {
+  fn from(n: MultiPlatformExecuteProcess) -> Self {
+    NodeKey::MultiPlatformExecuteProcess(Box::new(n))
+  }
+}
 
-impl WrappedNode for ExecuteProcess {
+impl WrappedNode for MultiPlatformExecuteProcess {
   type Item = ProcessResult;
 
   fn run(self, context: Context) -> NodeFuture<ProcessResult> {
     let request = self.0;
     let workunit_store = context.session.workunit_store();
-    context
+    if context
```

Comment:
I would rather have the `CommandRunner`s return an `Err` if there was no compatible request, than do this `if` check and call `unwrap()` in the individual `CommandRunner`s

Reply:
I agree that fits the rust paradigm better, but when designing this with @stuhood he was adamant that no incompatible request be "run" async. That is, that compatibility should be figured out synchronously.

Reply:
I think that this might be a misreading... this only calls unwrap on `Platform::current`: it returns an Err for other failures. So I think it's already doing what @illicitonion is suggesting?

Reply:
The local and remote command runners do .unwrap() the Result from extract_compatible_request, but only after this check so they are guaranteed to succeed.